DDOT_module/child_classes/ddot_marginal_analyzer_vanilla.py

```diff
@@ -0,0 +1,396 @@
+        averaging_normalization = int(2 ** (len(all_neighbors) - len(qubits_to_be_left)))
+
+        states_after_averaging = anf.register_names_qubits(range(len(qubits_to_be_left)),
+                                                           len(qubits_to_be_left), False)
+        averaged_dimension = list(matrices_cluster.values())[0].shape[0]
+
+        averaged_matrices_cluster = {
+            state: np.zeros((averaged_dimension, averaged_dimension), dtype=float) for state in
+            states_after_averaging}
+
+        qubits_to_be_averaged_over = list(set(all_neighbors).difference(set(qubits_to_be_left)))
+        qubits_to_be_averaged_over_mapped = [reversed_enumerated[q] for q in
+                                             qubits_to_be_averaged_over]
+
+        for neighbors_state, conditional_noise_matrix in matrices_cluster.items():
+            list_string_neighbors = list(copy.deepcopy(neighbors_state))
+
+            list_string_neighbors = list(np.delete(list_string_neighbors,
+                                                   qubits_to_be_averaged_over_mapped))
+
+            string_neighbors = ''.join(list_string_neighbors)
+
+            # TODO FBM: make sure this commented check is not needed anymore
+            # if string_neighbors == '' or string_neighbors:
+            #     try:
+            #         averaged_matrices_cluster['averaged'] +=
+            #         conditional_noise_matrix / averaging_normalization
+            #     except(KeyError):
+            #         averaged_matrices_cluster['averaged'] =
+            #         conditional_noise_matrix/averaging_normalization
+            # else:
+            averaged_matrices_cluster[
+                string_neighbors] += conditional_noise_matrix / averaging_normalization
+
+        return averaged_matrices_cluster
+
+    def _compute_noise_matrix_averaged(self,
+                                       subset: List[int]) -> np.ndarray:
+        """Noise matrix for subset of qubits, averaged over all other qubits
+           Defaultly takes data from self._marginals_dictionary. If data is not present, then it
+           calculates marginals for given subset
+           and updates the class's property self.marginals_dictionary
+        """
+
+        # TODO: Perhaps add possibility of using existing marginals for bigger subset that includes
+        # target subset
+
+        subset_key = 'q' + 'q'.join([str(s) for s in subset])
+        marginals_dictionary_ddot = self._marginals_dictionary
+
+        # TODO FBM: testing some issue with normalization
+        testing_method = False
+
+        # TODO FBM: refactor defaultdict usage here
+        marginal_dict_now = defaultdict(float)
+
+        for input_state_bitstring, dictionary_marginals_now in marginals_dictionary_ddot.items():
+            input_marginal = ''.join([input_state_bitstring[x] for x in subset])
+
+            if subset_key not in dictionary_marginals_now.keys():
```

Comment:
I believe `.keys()` is not needed.

Reply:
I know, but I prefer to leave it anyway to avoid confusion. (it would be confusing for me)